### PR TITLE
f64: radix-4 butterfly stage ButterflyComplexStage4 (Fixes #198)

### LINKED ISSUE #198: f64: radix-4 butterfly-stage variant (follow-up to #192; parts 1-2 shipped)
Follow-up to #192, from adopting v1.6.0 in [go-spectrogram](https://github.com/tphakala/go-spectrogram)'s SoX-compatible renderer. Both new f64 kernels are wired in and measured against the scalar code they replace. `ButterflyComplex` is a clear win and is now in use; `RealFFTUnpack` measured slower than the scalar loop and was backed out. Both results point at the same gap, so this is one issue rather than two.

## 1. `ButterflyComplex` needs a stage-level form

The kernel takes one contiguous run of pairs, so driving a full radix-2 decimation-in-time stage at span `s` over a transform of size `n` costs `n/(2s)` separate calls. The butterflies per stage are constant at `n/2`, so the call count explodes exactly where the runs get short, and per-call overhead ends up dominating the small-span stages.

Measured on an idle i7-1260P (AVX2+FMA), one complete radix-2 stage over `n = 1024` split-complex float64, kernel-driven versus the equivalent scalar Go loop:

| span | calls | via `ButterflyComplex` | scalar loop | speedup |
|------|-------|------------------------|-------------|---------|
| 4    | 128   | 1143 ns | 1519 ns | 1.3x |
| 8    | 64    | 647 ns  | 1423 ns | 2.2x |
| 16   | 32    | 389 ns  | 1444 ns | 3.7x |
| 32   | 16    | 265 ns  | 1690 ns | 6.4x |
| 64   | 8     | 192 ns  | 1366 ns | 7.1x |
| 128  | 4     | 162 ns  | 1360 ns | 8.4x |
| 256  | 2     | 148 ns  | 1327 ns | 9.0x |
| 512  | 1     | 140 ns  | 1367 ns | 9.8x |

The work is identical in every row. The whole spread is call overhead: roughly 8 ns per call, against about 1 ns of actual butterflies at span 4.

The practical consequence is that a consumer cannot simply use the kernel for the whole transform. go-spectrogram now runs a scalar radix-4 butterfly for the small spans and switches to `ButterflyComplex` from span 8 upward, because one scalar radix-4 stage does the work of two radix-2 stages for well under twice the cost, which beats the kernel until the runs are long enough. Half the stages of a decimation-in-time transform have span below `sqrt(n)`, so a good fraction of the butterflies stay scalar.

What would remove the trade entirely is a kernel that takes the stage rather than the block, something like:

```go
// ButterflyComplexStage applies one radix-2 decimation-in-time stage in place
// over split-complex data: for every block k in steps of 2*span, and every j in
// [0, span), the butterfly at (k+j, k+span+j) with twiddle tw[j].
func ButterflyComplexStage(re, im []float64, span int, twRe, twIm []float64)
```

Taking the whole stage lets the implementation choose its own vectorization axis: over `j` when the span is long, and over `k` (many short blocks at a fixed `j`) when it is short, which is the standard way small-span stages get vectorized and is not expressible through the current per-block API. On the numbers above, that would bring every stage to roughly the span-512 rate, which is a further 3x or so on the transform core beyond what v1.6.0 already delivers.

A radix-4 variant taking `span` and three twiddle arrays would be better still, since radix 4 does the work of two radix-2 passes with three quarters of the complex multiplies and half the passes over the scratch.

## 2. `RealFFTUnpack` needs a power-writing variant

`RealFFTUnpack` writes the complex half-spectrum. Consumers computing a spectrogram want `|X_k|^2`, so the unpack has to be followed by two more passes over the bins to square and sum. That extra traffic costs more than the vector lanes gain.

Measured on an idle i7-1260P at a fixed clock, `n = 8` runs through benchstat, full real-input power spectrum of one frame, `RealFFTUnpack` + `Mul` + `FMA` against the scalar loop it replaced:

```
                     │   scalar    │               unpack                │
                     │   sec/op    │    sec/op     vs base               │
PowerInto/nfft256      1.250µ ± 2%    1.739µ ± 0%  +39.12% (p=0.000 n=8)
PowerInto/nfft512      2.241µ ± 2%    2.571µ ± 6%  +14.70% (p=0.000 n=8)
PowerInto/nfft1024     5.534µ ± 4%    5.440µ ± 5%        ~ (p=0.487 n=8)
PowerInto/nfft2048     9.832µ ± 0%   10.231µ ± 3%   +4.06% (p=0.000 n=8)
geomean                3.514µ         3.971µ       +13.03%
```

The scalar loop wins because it emits power directly and reads bins `k` and `n-k` together, so one set of loads produces two output bins and nothing intermediate is ever written. So the kernel is correct and vectorized and still loses, purely on memory traffic.

A fused variant would flip that:

```go
// RealFFTPower writes |X_k|^2 for k in [1, n-1] given the half-size complex
// spectrum of a packed real transform, without materialising X.
func RealFFTPower(dst, zRe, zIm, twRe, twIm []float64)
```

This is the more useful entry point for any spectrogram, mel front end, or PSD consumer, which is most of what a real FFT gets used for; the complex output is the rarer need.

## Context

For scale, the renderer these were measured in currently spends about 55% of its time in the transform at its largest size, of which the butterfly is the great majority. v1.6.0's `ButterflyComplex` cut the full real-input power spectrum at nfft 2048 by roughly 20% on its own, on top of earlier scalar restructuring, and parity against the `sox` binary stayed at 100.000% exact with worst delta 0 on both AVX2 and NEON, so the kernel is bit-compatible in practice for this consumer.

### PR DESCRIPTION
**Pull request by tphakala**
Adds `f64.ButterflyComplexStage4`, a radix-4 decimation-in-time FFT butterfly stage that does the work of two radix-2 stages (at span `s` then `2*s`) in a single pass over the data. This closes #198.

One radix-4 stage combines the four sub-vectors of each block while they are still in registers, so it halves the passes over the array versus running two radix-2 stages. It ships a pure-Go reference (the source of truth), an AMD64 AVX+FMA kernel, and an ARM64 NEON kernel, mirroring the existing `ButterflyComplexStage`.

## Math

For each block `k` in steps of `4*span` and each `j` in `[0, span)`, with `x0..x3` at `(k+j, k+span+j, k+2*span+j, k+3*span+j)`:

```
t1 = x1*tw1[j]; t2 = x2*tw2[j]; t3 = x3*tw3[j]   (full complex multiplies)
a = x0 + t1; b = x0 - t1; c = t2 + t3; d = t2 - t3
y0 = a + c; y1 = b - i*d; y2 = a - c; y3 = b + i*d
```

The three twiddle slices are the powers of `w = exp(-2*pi*i/(4*span))`: `tw1 = w^(2j)`, `tw2 = w^j`, `tw3 = w^(3j)`. `x1` carries `w^(2j)` (not `w^j`) because a decimation-in-time input is radix-2 bit-reversed, which swaps sub-vectors 1 and 2. With this convention `tw1` is exactly the radix-2 span-`s` twiddle table and `tw2` is the first `s` entries of the radix-2 span-`2s` table, so a caller already holding the radix-2 tables reuses them. The `-i`/`+i` cross-adds hard-code the forward DFT direction. A test pins the identity that one radix-4 stage equals the two radix-2 stages it replaces.

## Kernels

AMD64 is AVX+FMA only (no AVX2): the `span == 1` block-axis path transposes a 4x4 float64 tile with `VUNPCKLPD`/`VUNPCKHPD`/`VPERM2F128` and splats its twiddles with the memory-source form of `VBROADCASTSD`; `span >= 2` vectorizes across `j`. ARM64 uses `.2D` NEON with `UZP1`/`UZP2`/`ZIP1`/`ZIP2` for the `span == 1` transpose; the new `.2D` arithmetic and permute encodings are hand-encoded `WORD`s whose decoded-mnemonic comments are cross-checked against `arm64asm`.

## Benchmarks

20 independent rounds per arm, one sweep per process, Mann-Whitney U. `ratio = TwoRadix2 / Stage4` (a value above 1 means the single radix-4 stage beats the two radix-2 stages it replaces). `FullCore` runs a complete transform: 5 radix-4 stages versus the 10 radix-2 stages that do the same work (plus a trailing radix-2 stage for the `2048` size).

AMD64 (AVX+FMA):

| span | ratio | p |
|-----:|------:|--:|
| 1 | 1.00 | 0.54 (tied) |
| 4 | 1.35 | 3e-8 |
| 16 | 1.35 | 2e-7 |
| 64 | 1.45 | 4e-8 |
| 256 | 1.48 | 3e-8 |
| FullCore1024 | 1.23 | 4.5e-7 |
| FullCore2048 | 1.24 | 3e-8 |

ARM64 (Cortex-A76, NEON):

| span | ratio | p |
|-----:|------:|--:|
| 1 | 1.61 | 3e-8 |
| 4 | 1.38 | 3e-8 |
| 16 | 1.28 | 3e-8 |
| 64 | 1.28 | 3e-8 |
| 256 | 1.28 | 3e-8 |
| FullCore1024 | 1.38 | 3e-8 |
| FullCore2048 | 1.32 | 3e-8 |

The full-transform speedup is the metric that matters for #198, and it holds on both architectures. A radix-4 schedule only ever uses spans 1, 4, 16, 64, ..., so spans 2 and 3 are intentionally left to the scalar tail rather than carrying dedicated vector paths. On AMD64 the span-1 stage is statistically tied with the two radix-2 calls in isolation but is not a regression, and the whole transform still wins because the later stages do; on ARM64 the 2-wide NEON path wins span 1 outright.

## Tests

Pinning test (one radix-4 stage == two radix-2 stages), parity against an independent complex128 reference, a full iterative Cooley-Tukey transform checked against a naive DFT, partial-block-untouched and degenerate-input no-op checks, an allocation-free assertion, and a differential fuzz target. All pass on AMD64 (native and with the SIMD path disabled to exercise the Go fallback) and on ARM64 hardware.

## Follow-ups (filed separately, not blockers)

The f32 sibling and adopting radix-4 inside `STFTPlan`.

## Summary by CodeRabbit

* **New Features**
  * Added a public float64 radix-4 complex FFT butterfly stage for in-place processing.
  * Supports AVX+FMA, ARM64 NEON, and portable Go execution paths.
  * Handles twiddle factors, partial blocks, span-based processing, and edge cases safely.

* **Tests**
  * Added comprehensive correctness, compatibility, fuzz, allocation, and benchmark coverage.
  * Documented the new FFT building block and its operation details.

### CODE REVIEW

File: f64/f64.go [hunk 1]
```diff
@@ -949,6 +949,84 @@ func ButterflyComplexStage(re, im []float64, span int, twRe, twIm []float64) {
...
+	tw1Re, tw1Im, tw2Re, tw2Im, tw3Re, tw3Im []float64) {
+	if span <= 0 ||
+		len(tw1Re) < span || len(tw1Im) < span ||
+		len(tw2Re) < span || len(tw2Im) < span ||
+		len(tw3Re) < span || len(tw3Im) < span {
+		return
+	}
+	n := min(len(re), len(im))
+	blockLen := butterflyStage4Radix * span
+	blocks := n / blockLen
```
[the remark this answers is not in the document]
Comment (tphakala) [reply f64/f64.go:1021]: Guarded by the length checks right above it. To reach the `blocks := n / blockLen` divide the call has to first pass `len(tw1Re) >= span` (plus the other five twiddle slices), and for `4*span` to wrap to 0 you'd need span >= 2^62, which no float64 slice can satisfy since Go caps slice length far below that. So an overflowing span already returns as a no-op at the twiddle-length guard and never reaches the divide. I also kept this consistent with the existing radix-2 ButterflyComplexStage, which computes `blockLen := butterflyStageRadix * span` the same way, rather than adding an unreachable guard here only.
